src/proxy/socks5.rs

```diff
@@ -229,3 +249,53 @@ async fn handle(
     });
     Ok(())
 }
+
+async fn dns_lookup(
+    resolver: Arc<dyn Resolver + Send + Sync>,
+    client_addr: SocketAddr,
+    hostname: &str,
+) -> Result<IpAddr, Error> {
+    fn new_message(name: Name, rr_type: RecordType) -> Message {
+        let mut msg = Message::new();
+        msg.set_id(rand::random());
+        msg.set_message_type(MessageType::Query);
+        msg.set_recursion_desired(true);
+        msg.add_query(Query::query(name, rr_type));
+        msg
+    }
+    /// Converts the given [Message] into a server-side [Request] with dummy values for
+    /// the client IP and protocol.
+    fn server_request(msg: &Message, client_addr: SocketAddr, protocol: Protocol) -> Request {
+        let wire_bytes = msg.to_vec().unwrap();
+        let msg_request = MessageRequest::from_bytes(&wire_bytes).unwrap();
+        Request::new(msg_request, client_addr, protocol)
+    }
+
+    /// Creates a A-record [Request] for the given name.
+    fn a_request(name: Name, client_addr: SocketAddr, protocol: Protocol) -> Request {
+        server_request(&new_message(name, RecordType::A), client_addr, protocol)
+    }
+
+    /// Creates a AAAA-record [Request] for the given name.
+    fn aaaa_request(name: Name, client_addr: SocketAddr, protocol: Protocol) -> Request {
+        server_request(&new_message(name, RecordType::AAAA), client_addr, protocol)
+    }
+
+    // TODO: do we need to do the search?
+    let name = Name::from_utf8(hostname)?;
+
+    // TODO: we probably want to race them or something. Is there something higher level that can handle this for us?
```

Comment:
Should probably be:

- IF client_addr == ipv6

- build both, `tokio::select`

yea

Reply:
Actually AFAICT the go dns code waits for both to fail, so maybe we should `join` and not `select`